### PR TITLE
Make http connection persistent

### PR DESCRIPTION
**Pull request by Mohammad-Haris**
Curl should use the already established connection, if there is one.
Closes Issue #22.

Helpful link regarding test: https://curl.se/libcurl/c/CURLINFO_NUM_CONNECTS.html

### CODE REVIEW

File: src/Unirest/Request.php [hunk 1]
```diff
@@ -22,8 +21,12 @@ class Request
     private static $overrideRetryForNextRequest = OverrideRetry::USE_GLOBAL_SETTINGS;
     private static $verifyPeer = true;
     private static $verifyHost = true;
+    private static $connectionTimeout = 300;
+    private static $defaultHeaders = array(
+        'Connection: Keep-Alive'
```
Comment (asadali214) [inline src/Unirest/Request.php]: In PHP array are key-value pairs, and for default headers (key = headerName,  value = headerValue), if you try to send headers like this it will be received on server with numeric keys: 
image

File: src/Unirest/Request.php [hunk 2]
```diff
@@ -34,13 +37,15 @@ class Request
...
         'type' => CURLPROXY_HTTP,
-        'auth' => array (
+        'auth' => array(
             'user' => '',
             'pass' => '',
             'method' => CURLAUTH_BASIC
         )
     );
 
+    public static $prevCallSuccessfulConnects;
```
Comment (asadali214) [inline src/Unirest/Request.php]: public variable means we are exposing it our sdk user, avoid adding public functions or variables in these libraries moreover initiate this with an int value `0`

File: src/Unirest/Request.php [hunk 3]
```diff
@@ -496,34 +514,47 @@ public static function buildHTTPCurlQuery($data, $parent = false)
...
      */
     public static function send($method, $url, $body = null, $headers = array(), $username = null, $password = null)
     {
-        self::$handle = curl_init();
+        if (self::$handle == null) {
+            self::$handle = curl_init();
+        }
+
+        else {
+            self::$prevCallSuccessfulConnects = curl_getinfo(self::$handle, CURLINFO_NUM_CONNECTS);
```
Comment (asadali214) [inline src/Unirest/Request.php]: Instead of replacing the current value keep a variable with a count of opened connections and increment that count using `curl_getinfo(self::$handle, CURLINFO_NUM_CONNECTS)` at the end of this method i.e. after the response is received
Then write a test case that checks opened connections count after sending multiple API calls to same and different Domains
This count should be 2 if 2 different Domains are used and so on.

File: src/Unirest/Request.php [hunk 4]
```diff
@@ -474,13 +490,15 @@ public static function buildHTTPCurlQuery($data, $parent = false)
         foreach ($data as $key => $value) {
             if ($parent) {
                 $new_key = sprintf('%s[%s]', $parent, $key);
-            } else {
+            }
+            else {
```
Comment (asadali214) [inline src/Unirest/Request.php]: undo all these else-block moves and refactoring changes that are done in this PR

File: src/Unirest/Request.php [hunk 5]
```diff
@@ -496,34 +514,47 @@ public static function buildHTTPCurlQuery($data, $parent = false)
...
                 $url .= '?';
             }
 
             $url .= urldecode(http_build_query(self::buildHTTPCurlQuery($body)));
         }
 
+        array_push(self::$defaultHeaders, "Keep-Alive: timeout=" . self::$connectionTimeout);
```
Comment (asadali214) [inline src/Unirest/Request.php]: This is causing a big issue in the headers of an endpoint call, each time a call is made, this line inserts a new value in the default headers, causing the array size to grow over the time

File: src/Unirest/Request.php [hunk 6]
```diff
@@ -563,27 +594,27 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+                CURLOPT_USERPWD => self::$auth['user'] . ':' . self::$auth['pass']
             ));
         }
 
         if (self::$proxy['address'] !== false) {
             curl_setopt_array(self::$handle, array(
-                CURLOPT_PROXYTYPE       => self::$proxy['type'],
-                CURLOPT_PROXY           => self::$proxy['address'],
-                CURLOPT_PROXYPORT       => self::$proxy['port'],
+                CURLOPT_PROXYTYPE => self::$proxy['type'],
```
Comment (asadali214) [inline src/Unirest/Request.php]: We have kept these arrows `=>` above and below each other, to keep the code readable, kindly revert these changes

File: src/Unirest/Request.php [hunk 7]
```diff
@@ -563,27 +594,27 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
-                CURLOPT_PROXYUSERPWD    => self::$proxy['auth']['user'] . ':' . self::$proxy['auth']['pass']
+                CURLOPT_PROXYAUTH => self::$proxy['auth']['method'],
+                CURLOPT_PROXYUSERPWD => self::$proxy['auth']['user'] . ':' . self::$proxy['auth']['pass']
             ));
         }
 
-        $retryCount      = 0;                           // current retry count
-        $waitTime        = 0.0;                         // wait time in secs before current api call
+        $retryCount = 0;                           // current retry count
+        $waitTime = 0.0;                         // wait time in secs before current api call
```
Comment (asadali214) [inline src/Unirest/Request.php]: similar to the `=>` we wese also keeping `=` above and below each other, you need to revert these changes too. Its a standard way of writing code in PHP to keep it readable

File: tests/Unirest/RequestTest.php [hunk 8]
```diff
@@ -83,6 +83,33 @@ public function testDefaultHeader()
         $this->assertFalse(property_exists($response->body->headers, 'hello'));
     }
 
+    public function testConnectionReuse()
```
Comment (asadali214) [inline tests/Unirest/RequestTest.php:88]: After you implement the increasing count of connections this test case must be updated

File: tests/Unirest/RequestTest.php [hunk 9]
```diff
@@ -83,6 +83,33 @@ public function testDefaultHeader()
...
+            $_headers = [
+                'Accept' => 'application/json'
+            ];
+
+            Request::get($url, $_headers);
+
+            sleep(1);
+        }
+
+        $this->assertEquals(1, Request::$prevCallSuccessfulConnects);
```
Comment (asadali214) [inline tests/Unirest/RequestTest.php]: Instead of accessing the public property like this, make it protected then add a child class to access the parent's property using a public getter in the child class, Just like you did in JsonMapper's test case

File: tests/Unirest/RequestTest.php [hunk 10]
```diff
@@ -83,6 +83,33 @@ public function testDefaultHeader()
         $this->assertFalse(property_exists($response->body->headers, 'hello'));
     }
 
+    public function testConnectionReuse()
+    {
+        for ($i = 0; $i < 2; $i++) {
+            $url = "http://httpbin.org/get";
```
Comment (asadali214) [inline tests/Unirest/RequestTest.php]: $url and $header variables can be declared once and reused

**Review by asadali214 — CHANGES_REQUESTED** [review]
I have completed 1st review of your PR, there are alot of refactoring related changes like removing spaces etc, kindly remove all such changes

File: src/Unirest/Request.php (hunk 1 above)
Comment (Mohammad-Haris) [reply src/Unirest/Request.php]: Needs some discussion.

File: src/Unirest/Request.php (hunk 3 above)
Comment (Mohammad-Haris) [reply src/Unirest/Request.php]: Does it add something to the test we already have? It checks the same thing in a different way. In fact checking new connection count against 0 after previously having the same connection is a more direct way of checking if the connection was re-used or not.

File: tests/Unirest/RequestTest.php [hunk 11]
```diff
@@ -83,6 +84,126 @@ public function testDefaultHeader()
         $this->assertFalse(property_exists($response->body->headers, 'hello'));
     }
 
+    public function testConnectionReuse()
+    {
+        RequestChild::resetHandleAndPrevConnects();
```
Comment (asadali214) [inline tests/Unirest/RequestTest.php]: Very good handling for connections reseting

File: src/Unirest/Request.php [hunk 12]
```diff
@@ -22,6 +21,10 @@ class Request
     private static $overrideRetryForNextRequest = OverrideRetry::USE_GLOBAL_SETTINGS;
     private static $verifyPeer = true;
     private static $verifyHost = true;
+    private static $connectionTimeout = 300;
```
Comment (asadali214) [inline src/Unirest/Request.php]: I have updated the PR with a few changes, after an in-depth review of the functionality added but I still couldn't understand the use of Timeout header, is it relevant with connections header?

**Review by asadali214 — APPROVED** [review]
Approved, and good work!
